pkg/controllers/ragengine_status.go

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package controllers
+
+import (
+	"context"
+
+	kaitov1alpha1 "github.com/azure/kaito/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (c *RAGEngineReconciler) updateRAGEngineStatus(ctx context.Context, name *client.ObjectKey, condition *metav1.Condition, workerNodes []string) error {
+	return retry.OnError(retry.DefaultRetry,
+		func(err error) bool {
+			return apierrors.IsServiceUnavailable(err) || apierrors.IsServerTimeout(err) || apierrors.IsTooManyRequests(err)
+		},
+		func() error {
+			// Read the latest version to avoid update conflict.
+			ragObj := &kaitov1alpha1.RAGEngine{}
+			if err := c.Client.Get(ctx, *name, ragObj); err != nil {
+				if !errors.IsNotFound(err) {
+					return err
+				}
+				return nil
+			}
+			if condition != nil {
+				meta.SetStatusCondition(&ragObj.Status.Conditions, *condition)
+			}
+			if workerNodes != nil {
+				ragObj.Status.WorkerNodes = workerNodes
+			}
+			return c.Client.Status().Update(ctx, ragObj)
+		})
+}
+
+func (c *RAGEngineReconciler) updateStatusConditionIfNotMatch(ctx context.Context, ragObj *kaitov1alpha1.RAGEngine, cType kaitov1alpha1.ConditionType,
```

Comment:
the same logic as https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_status.go#L47